### PR TITLE
Remove last traces of python 2 special handling

### PR DESCRIPTION
**Pull request by ZedThree**
PR into #322 removing last few bits of `six` and python 2 special handling

### CODE REVIEW

**Comment by arporter** [comment]
Hi @ZedThree, if you could fix the conflicts on this one and then I'll review it... Thanks :-)

**Comment by ZedThree** [comment]
@arporter Good to go :)

**Review by arporter — CHANGES_REQUESTED** [review]
Many thanks for taking this on Peter - it's a big improvement.
GHA and coverage is all fine.
Documentation builds and looks fine.
There's one method that still talks quite a lot about Py2 and Py3.
There's also one test that you've removed completely and I'm not sure why.
Apart from that, there are just a few places where I've taken the opportunity to ask for fixes to associated pylint warnings. (I tried to keep this to a minimum because otherwise you'd be here all year given the state of some of the existing code.)

File: src/fparser/common/tests/test_readfortran.py [hunk 1]
```diff
@@ -374,7 +374,7 @@ def check_include_works(fortran_filename, fortran_code, include_info,
     file specified by 'fortran_filename'.
     :param include_info: a list of 2-tuples each with an include \
     filename as a string followed by include code as a string.
-    :type include_info: list of (str, str)
+    :type include_info: list of str
```
[the remark this answers is not in the document]
Comment (ZedThree) [reply src/fparser/common/tests/test_readfortran.py:377]: Sure, will do. Future project might be to migrate to type hinting?

File: src/fparser/two/tests/fortran2003/test_entity_decl_r504.py [hunk 2]
```diff
@@ -40,27 +40,6 @@
 from fparser.two.Fortran2003 import Entity_Decl, Name
 
 import pytest
-# TODO #307 remove this once we drop Python 2
-import six
-
-
-def test_entity_decl_repr():
```
[the remark this answers is not in the document]
Comment (ZedThree) [reply src/fparser/two/tests/fortran2003/test_entity_decl_r504.py:47]: Oops, sorry, moving too fast! Will revert!

File: src/fparser/common/base_classes.py [hunk 3]
```diff
@@ -243,9 +242,8 @@ def set_init(self, expr):
         return
 
     def set_dimension(self, dims):
-        import six
         dims = [tuple(dim.split(':')) for dim in dims]
-        dims = [tuple(map(six.text_type.strip, dim)) for dim in dims]
+        dims = [tuple(map(str.strip, dim)) for dim in dims]
```
[the remark this answers is not in the document]
Comment (ZedThree) [reply src/fparser/common/base_classes.py:246]: I've now done this for all files in `common` and `two` that matched the regexp `^\ {4,8}return$` and then manual checking

**Comment by ZedThree** [comment]
One other change that I've not done but just seen is use of `match = staticmethod(match)` which could be converted to be a decorator on the method instead. This is a pretty mechanical fix (just taken me a couple of seconds to do with an Emacs macro), would you like that in here too? That's 135 lines

**Comment by arporter** [comment]
Hi @ZedThree, is this ready for review again?

File: src/fparser/common/tests/test_readfortran.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/common/tests/test_readfortran.py]: Yes, ultimately that would be good.

File: src/fparser/two/Fortran2003.py [hunk 4]
```diff
@@ -2951,7 +2947,6 @@ def match(string):
             return '=>', Null_Init(string[2:].lstrip())
         if string.startswith('='):
             return '=', Initialization_Expr(string[1:].lstrip())
-        return
```
Comment (arporter) [inline src/fparser/two/Fortran2003.py:2955]: return None pls.

File: src/fparser/two/Fortran2003.py [hunk 5]
```diff
@@ -4643,7 +4635,6 @@ def match(string):
                 obj = None
             if obj is not None:
                 return obj
-        return
```
Comment (arporter) [inline src/fparser/two/Fortran2003.py:4647]: `return None` please.

File: src/fparser/two/Fortran2003.py [hunk 6]
```diff
@@ -8720,7 +8711,6 @@ def match(string):
             if not line:
                 return c, None, lst, None
             return c, Char_Literal_Constant(line), lst, None
-        return
```
Comment (arporter) [inline src/fparser/two/Fortran2003.py:8724]: `return None`

File: src/fparser/two/utils.py [hunk 7]
```diff
@@ -1300,7 +1299,6 @@ def match(pattern, string):
             return
         if pattern.match(string):
             return string,
-        return
```
Comment (arporter) [inline src/fparser/two/utils.py:1303]: return None pls.

File: src/fparser/two/utils.py [hunk 8]
```diff
@@ -1378,7 +1376,7 @@ def match(my_pattern, string):
         except AttributeError:
             raise InternalError(
                 "Supplied pattern should be a list, tuple, str or regular "
```
Comment (arporter) [inline src/fparser/two/utils.py:1378]: Our standard is to prefix all of the sub-strings with `f` for consistency.

File: src/fparser/common/utils.py [hunk 9]
```diff
@@ -269,7 +269,6 @@ def get_module_file(name, directory, _cache={}):
         if module_in_file(name, fn):
             _cache[name] = fn
             return fn
-    return
```
[the remark this answers is not in the document]
Comment (ZedThree) [reply src/fparser/common/utils.py:272]: Sure, will do. To be clear, this is equivalent, but I agree that `return None` is more explicit :+1:

**Review by arporter — CHANGES_REQUESTED** [review]
Almost there apart from those places where a `return None` is needed :-)
Please note that our way of working on GitHub is for the reviewer to mark conversations as 'resolved' as that makes it easier for them to ensure that each has been done to their satisfaction.

File: src/fparser/common/base_classes.py [hunk 10]
```diff
@@ -734,7 +715,7 @@ def get_provides(self):
         Returns dictonary containing statements that block provides or None
         when N/A.
         """
-        return
+        pass
```
[the remark this answers is not in the document]
Comment (ZedThree) [reply src/fparser/common/base_classes.py:718]: True, but `pass` is probably more explicit that the method hasn't been implemented? Or perhaps even `raise NotImplementedError` if you want to ensure subclasses implement it

File: src/fparser/common/base_classes.py [hunk 11]
```diff
@@ -926,7 +904,6 @@ def process_subitem(self, item):
                 return r
 
         self.handle_unknown_item(item)
-        return
```
[the remark this answers is not in the document]
Comment (ZedThree) [reply src/fparser/common/base_classes.py:929]: `self.handle_unknown_item` will unconditionally raise an exception

File: src/fparser/common/base_classes.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/common/base_classes.py:718]: On coming back to this I think `return None` is the appropriate thing to do - this then lines up with the docstring.

File: src/fparser/common/base_classes.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/common/base_classes.py:929]: Aha. Please could you add a comment to that effect before `self.handle_unknown_item(item)`.

[the remark this answers is not in the document]
Comment (ZedThree) [reply src/fparser/common/base_classes.py:929]: Maybe instead we change the function name to `raise_unknown_item_error` or similar?

File: src/fparser/common/base_classes.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (arporter) [reply src/fparser/common/base_classes.py:718]: Please could you change this to `return None`.

**Review by arporter — APPROVED** [review]
All good now thanks. Will proceed to merge once CI is complete (and happy).

**Comment by arporter** [comment]
The missed lines are all statements where `return` has been replaced by `return None` in `Fortran2003.py`. These weren't covered previously and I'm not going to insist they be covered now.

**Pull request merged**